rules/prefer-global-this.js

```diff
@@ -0,0 +1,259 @@
+	'name',
+	'postMessage',
+	'onconnect',
+]);
+
+/**
+Report the node with a message.
+
+@param {import('eslint').Rule.RuleContext} context
+@param {import('estree').Node} node
+@param {string} value
+*/
+function report(context, node, value) {
+	const fix = (/** @type {import('eslint').Rule.RuleFixer} fixer */ fixer) => fixer.replaceText(node, 'globalThis');
+	context.report({
+		node,
+		messageId: MESSAGE_ID_ERROR,
+		data: {value, replacement: 'globalThis'},
+		fix,
+		suggest: [{messageId: MESSAGE_ID_SUGGESTION, data: {value, replacement: 'globalThis'}, fix}],
+	});
+}
+
+/**
+Handle nodes and check if they should be reported.
+
+@param {import('eslint').Rule.RuleContext} context
+@param {import('estree').Node | Array<import('estree').Node>} nodes
+*/
+function handleNodes(context, nodes) {
+	if (!Array.isArray(nodes)) {
+		nodes = [nodes];
+	}
+
+	for (const node of nodes) {
+		if (node && node.type === 'Identifier' && globalIdentifier.has(node.name)) {
+			const variable = findVariableInScope(context.sourceCode.getScope(node), node.name);
+			if (!variable) {
+				report(context, node, node.name);
+			}
+		}
+	}
+}
+
+/**
+Check if the node is a window-specific API.
+
+@param {import('estree').MemberExpression} node
+@returns {boolean}
+*/
+const isWindowSpecificAPI = node => {
+	if (node.type !== 'MemberExpression') {
+		return false;
+	}
+
+	if (node.object.name !== 'window' || node.property.type !== 'Identifier') {
+		return false;
+	}
+
+	if (windowSpecificAPIs.has(node.property.name)) {
+		if (['addEventListener', 'removeEventListener', 'dispatchEvent'].includes(node.property.name) && node.parent && node.parent.type === 'CallExpression') {
+			const argument = node.parent.arguments[0];
+			return argument && argument.type === 'Literal' && windowSpecificEvents.has(argument.value);
+		}
+
+		return true;
+	}
+
+	return false;
+};
+
+/**
+Check if the node is a web worker specific API.
+
+@param {import('estree').MemberExpression} node
+@returns {boolean}
+*/
+const isWebWorkerSpecificAPI = node => node.type === 'MemberExpression' && node.object.name === 'self' && node.property.type === 'Identifier' && webWorkerSpecificAPIs.has(node.property.name);
+
+/** @param {import('eslint').Rule.RuleContext} context */
+const create = context => {
+	const tracker = new GlobalReferenceTracker({
+		objects: globalIdentifier,
+		handle(reference) {
+			const {node} = reference;
+
+			if (node.type === 'Identifier') {
+				if (isWindowSpecificAPI(node.parent) || isWebWorkerSpecificAPI(node.parent)) {
+					return;
+				}
+
+				report(context, node, node.name);
+			}
+		},
+	});
+
+	return {
+		...tracker.createListeners(context),
+		/** @param {import('estree').AssignmentExpression} node */
+		AssignmentExpression(node) {
+			handleNodes(context, [node.left]);
+		},
+		/** @param {import('estree').UpdateExpression} node */
+		UpdateExpression(node) {
```

Comment:
According to https://github.com/sindresorhus/eslint-plugin-unicorn/blob/77f32e5a6b2df542cf50dfbd371054f2cd8ce2d6/rules/utils/is-left-hand-side.js#L3, 
 Looks like we are missing `AssignmentPattern`
, `ArrayPattern`, `ObjectPattern` and `UnaryExpression`(`delete window.xxx`, but maybe no need to care.)

Reply:
`delete window.foo` is a `MemberExpression` that we already tested.

```md
## invalid(27): delete window.foo

> Input

    `␊
      1 | delete window.foo␊
    `

> Output

    `␊
      1 | delete globalThis.foo␊
    `

> Error 1/1

    `␊
    > 1 | delete window.foo␊
        |        ^^^^^^ Prefer \`globalThis\` over \`window\`.␊
    `
```

Reply:
`ObjectPattern` and `ArrayPattern` are necessary to test.
```js
// ObjectPattern
const { window } = obj

// ArrayPattern
const [window] = obj
```

### AssignmentPattern

```markdown
## invalid(48): function greet({ name = window.foo }) {}

> Input

    `␊
      1 | function greet({ name = window.foo }) {}␊
    `

> Output

    `␊
      1 | function greet({ name = globalThis.foo }) {}␊
    `

> Error 1/1

    `␊
    > 1 | function greet({ name = window.foo }) {}␊
        |                         ^^^^^^ Prefer \`globalThis\` over \`window\`.␊
    `

```

Reply:
Can you add test for 

`({ foo = window.foo } =  {});`

and
`[window.foo] = []`;

Reply:
These are invalid syntax

```
unknown: Unexpected token, expected "," (1:28)

> 1 | function greet({ foo: window.foo } =  {}) {}
    |                             ^
```

```
unknown: Unexpected token, expected "," (1:22)

> 1 | function greet([window.foo] = []) {}
    |                       ^
```

Reply:
```markdown

## invalid(49): ({ foo: window.foo } = {})

> Input

    `␊
      1 | ({ foo: window.foo } =  {})␊
    `

> Output

    `␊
      1 | ({ foo: globalThis.foo } =  {})␊
    `

> Error 1/1

    `␊
    > 1 | ({ foo: window.foo } =  {})␊
        |         ^^^^^^ Prefer \`globalThis\` over \`window\`.␊
    `

## invalid(50): [window.foo] = []

> Input

    `␊
      1 | [window.foo] = []␊
    `

> Output

    `␊
      1 | [globalThis.foo] = []␊
    `

> Error 1/1

    `␊
    > 1 | [window.foo] = []␊
        |  ^^^^^^ Prefer \`globalThis\` over \`window\`.␊
    `

```